### PR TITLE
First pass integrating component spec - stubs complete.

### PR DESCRIPTION
**Pull request by EvanKomp**

### CODE REVIEW

File: gandy/models/models.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    ## to contain dictionary of callable metric classes from the metrics module
+    metrics = {} #gandy.metrics.metric_codex
+    """Available metrics defined in gandy.metrics"""
+    
+    def __init__(self, 
+                 xshape: Tuple[int], 
+                 yshape: Tuple[int], 
+                 **kwargs):
+        ## pseudocode
+        #. assert inputs
```
Comment (EvanKomp) [inline gandy/models/models.py]: Is this needed with typing?

File: gandy/models/models.py [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+        """
+        ## pseudocode
+        #. raise not implimented 
+        #. model is None
+        return model
+    
+    def train(self,
+              Xs: Iterable,
+              Ys: Iterable, 
+              session: str = None,
```
Comment (EvanKomp) [inline gandy/models/models.py:134]: add metric keyword argument

File: gandy/models/models.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        """
+        ## psuedocode
+        #. raise not implimented
+        #. set pred, unc to None
+        return predictions, uncertainties
+    
+    def score(self, 
+              Xs: Iterable,
+              Ys: Iterable,
+              metric: Union[str, Callable],
```
Comment (EvanKomp) [inline gandy/models/models.py:260]: give a default value later

File: gandy/models/models.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+            float:
+                Total score according to metric.
+            ndarray:
+                Score array for each prediction.
+        """
+        ## pseudocode
+        #. if statement to get metric object from metrics or specified
+        #. else raise undefined metric
+        #. check data
+        #. predictions, uncertainties = execute self.predict on Xs
```
Comment (EvanKomp) [inline gandy/models/models.py]: _predict, not predict

File: gandy/models/models.py [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+        #. pass predictions, uncertainties to metric get back costs
+        return cost, costs
```
Comment (EvanKomp) [inline gandy/models/models.py]: call loss. metric_value and metric_values instead

File: gandy/models/models.py [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+        
+    def save(self,
```
Comment (EvanKomp) [inline gandy/models/models.py:292]: Can we make this a child defined method instead eg. predict. Raise not implemented in parent

File: gandy/models/models.py [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+        ## print message about deleting model, build needs to be ran
+        return
+    
+    @property
+    def xshape(self):
+        """tuple of int: shape of example features"""
+        return self._xshape
+    
+    @xshape.setter(self, new_xshape):
+        ## test new shape, delete model
```
Comment (EvanKomp) [inline gandy/models/models.py:385]: where is the definition

File: gandy/models/models.py [hunk 8]
```diff
@@ -0,0 +1,358 @@
...
+        self._xshape = new_xshape
+        return
+    
+    @property
+    def yshape(self):
+        """tuple of int: shape of example label"""
+        return self._yshape
+    
+    @yshape.setter(self, new_yshape):
+        ## test new shape, delete model
```
Comment (EvanKomp) [inline gandy/models/models.py:396]: dude define the method

File: gandy/models/hypersearch.py [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+## imports
+from typing import Tuple, Iterable, Any, Object, Type, List
+
+import optuna
+import numpy
+
+import gandy.models.models
+
+## Typing
+Model = Type[import gandy.models.models.UncertaintyModel]
```
Comment (EvanKomp) [inline gandy/models/hypersearch.py]: import?

File: gandy/models/hypersearch.py [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+            float: 
+                The score of this trial, the quantity to minimize.
+        """
+        ## pseudocode
+        #. sample hypparams for this trial
+        #. depending on val_data, k, or val_frac
+        #.    split data based on above
+        #.    for each session, execute instances
+        #.    check for prune
+        return score
```
Comment (EvanKomp) [inline gandy/models/hypersearch.py]: score makes it seem good but we are trying to minimize, loss?

File: gandy/models/hypersearch.py [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+                
+        Returns:
+            float: 
+                The score of the model on the validation data.
+        """
+        ## pseudocode
+        #. construct model with hyparms and self kwargs
+        #. train model with hyparms and self kwargs
+        #. score model with self kwargs
+        return single_score
```
Comment (EvanKomp) [inline gandy/models/hypersearch.py]: loss not score

File: gandy/models/hypersearch.py [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+        #. train model with best_params training and kwargs
+        #. set self.best_model
+        return best_model
+        
+    @property
+    def search_space(self):
+        """dict: hyperparameter name to search space parirings"""
+        return self._search_space
+    
+    @search_space.setter(self, new_search_space):
```
Comment (EvanKomp) [inline gandy/models/hypersearch.py]: define the method

**Review by EvanKomp — COMMENTED** [review]
I can't request changes but you should make them...

File: gandy/models/bnns.py [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    in the neural network, the Bayesian approach learns weight distributions
+    - from which we can sample to produce an output for a given input - to
+    encode weight uncertainty.
+
+    Thank you to
+    https://keras.io/examples/keras_recipes/bayesian_neural_networks/
+    for a guide to implementing a BNN with Keras.
+    '''
+
+    def create_model_inputs(self, feature_names):
```
Comment (stetef) [inline gandy/models/bnns.py]: Added typing

File: gandy/quality_est/metrics.py [hunk 14]
```diff
@@ -0,0 +1,122 @@
...
+class RMSE(Metric):
+    '''
+
+    Root Mean Squared Error class which defines the structure used for 
+    computing the RMSE between the passed in datasets. Inherets the
+    properties of the parent class Metrics.
+    
+    '''
+
+    def calculate(**kwargs):
```
Comment (kmosko) [inline gandy/quality_est/metrics.py]: Add self keyword to methods within metric children classes

File: gandy/quality_est/metrics.py [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+    ''' Method that defines the mathematical formula necessary to compute the MSE. 
+        
+        Arguments:
+        
+        **kwargs - necessary keyword arguments to be passed into <calculate> method
+
+        Returns:   
+
```
Comment (kmosko) [inline gandy/quality_est/metrics.py:81]: Return final metric value and list of individual metrics for each prediction

File: gandy/quality_est/metrics.py [hunk 16]
```diff
@@ -0,0 +1,122 @@
...
+            real - data of real values that you want to compare the uncertainty model ouput to (eg. experimental data)  
+                
+                type == ndarray
+        
+        '''
+        
+        ## pseudocode 
+        
+        #  set self.args
+    return
```
Comment (kmosko) [inline gandy/quality_est/metrics.py]: Adjust tab for return statements

File: gandy/quality_est/metrics.py [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+    '''
+
+    F1 score class which defines the structure used for 
+    computing the F1 score between the passed in datasets. Inherets the
+    properties of the parent class Metrics.
+    
+    '''
+
+    def calculate(**kwargs):
+    ''' Method that defines the mathematical formula necessary to compute the RMSE.
```
Comment (EvanKomp) [inline gandy/quality_est/metrics.py]: Docstring not correct

File: gandy/quality_est/metrics.py [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+                type == ndarray
+            
+            real - data of real values that you want to compare the uncertainty model ouput to (eg. experimental data)  
+                
+                type == ndarray
+        
+        '''
+        
+        ## pseudocode 
+
```
Comment (EvanKomp) [inline gandy/quality_est/metrics.py]: call calculate in init

File: gandy/quality_est/metrics.py [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+                type == ndarray
+        
+        '''
+        
+        ## pseudocode 
+        
+        #  set self.args
+    return 
+
+
```
Comment (EvanKomp) [inline gandy/quality_est/metrics.py]: add a calculate method to parent that does nothing

File: gandy/quality_est/metrics.py [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+        **kwargs - necessary keyword arguments to be passed into <calculate> method
+
+        Returns:   
+
+            MSE_value - value of the Mean Squared error computed 
+
+                type == float
+     '''
+
+    return MSE_value
```
Comment (EvanKomp) [inline gandy/quality_est/metrics.py]: save the value insead of returning maybe

File: gandy/quality_est/metrics.py [hunk 21]
```diff
@@ -0,0 +1,122 @@
+## imports
```
Comment (EvanKomp) [inline gandy/quality_est/metrics.py]: add a docstring to the module

**Review by EvanKomp — COMMENTED** [review]
Nice work kyle, a few functional things.

File: gandy/models/gans.py [hunk 22]
```diff
@@ -0,0 +1,191 @@
...
+    # overridden method from UncertaintyModel class
+    def _build(self, **kwargs):
+        '''
+        Construct the model
+        '''
+        # do something like:
+        # self.create_generator(**kwargs)
+        # self.create_discriminator(**kwargs)
+        # self.n_classes = self.yshape
+        return None
```
Comment (EvanKomp) [inline gandy/models/gans.py]: return model for self.build

**Review by stetef — CHANGES_REQUESTED** [review]
Make appropriate changes (see comments)

**Review by stetef — APPROVED** [review]

**Review by kmosko — APPROVED** [review]
Changes look good!
